llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -4752,6 +4752,19 @@ unsigned SelectionDAG::ComputeNumSignBits(SDValue Op, const APInt &DemandedElts,
         (VTBits - SignBitsOp0 + 1) + (VTBits - SignBitsOp1 + 1);
     return OutValidBits > VTBits ? 1 : VTBits - OutValidBits + 1;
   }
+  case ISD::AVGCEILS:
+  case ISD::AVGFLOORS:
+    // Treat the AVG nodes like ADD/SUB nodes as that's the only stage that
+    // can lose sign bits:
+    // avgceils(lhs, rhs) -> sub(or(lhs,rhs),ashr(xor(lhs,rhs),1))
+    // avgfloors(lhs, rhs) -> add(and(lhs,rhs),ashr(xor(lhs,rhs),1))
+    Tmp = ComputeNumSignBits(Op.getOperand(0), DemandedElts, Depth + 1);
+    if (Tmp == 1)
+      break; // Early out.
+    Tmp2 = ComputeNumSignBits(Op.getOperand(1), DemandedElts, Depth + 1);
+    if (Tmp2 == 1)
+      break; // Early out.
+    return std::min(Tmp, Tmp2) - 1;
```

Comment:
Intuitively I would hope that we don't need the `- 1` here but I'm not sure. Do we have any exhaustive tests for NumSignBits calculations?

Reply:
No, I'm starting to wonder if we should have some form of SignBits equivalent to KnownBits so we can share more valuetracking code - what do you think?

Reply:
Not exhaustive, but random fuzzing suggests we can use `std::min(Tmp,Tmp2)` - https://godbolt.org/z/Trx1986rj

Reply:
> No, I'm starting to wonder if we should have some form of SignBits equivalent to KnownBits so we can share more valuetracking code - what do you think?

We could create something similar to KnownBits that calculates numsignbits for an operation based on the numsignbits of its inputs, but I'm not sure how useful that would be. My hunch is that most of the useful stuff that happens in numsignbits calculations relies on more information than just the numsignbits of the inputs.